File: frontend/src/app/core/services/realtime-transport.service.ts
import { Injectable, computed, signal } from '@angular/core';
import { Subject } from 'rxjs';
import { ChatApiService, RealtimeSocket } from './chat-api.service';

export type RealtimeTransportMode = 'socket' | 'sse' | 'polling';

const POLL_INTERVAL_MS = 15000;
const STREAM_RETRY_MS = 5000;
const SOCKET_RETRY_MS = 3500;
const SOCKET_FALLBACK_TO_SSE_DELAY_MS = 1800;
const SOCKET_MAX_FAILURES_BEFORE_COOLDOWN = 3;
const SOCKET_FAILURE_COOLDOWN_MS = 5 * 60 * 1000;
const SOCKET_ACK_TIMEOUT_MS = 6000;

@Injectable({ providedIn: 'root' })
export class RealtimeTransportService {
  /** Current transport mode. */
  readonly transportMode = signal<RealtimeTransportMode>('polling');

  /** Human-readable label for the current transport mode. */
  readonly transportLabel = computed(() => {
    const mode = this.transportMode();
    if (mode === 'socket') return 'Socket';
    if (mode === 'sse') return 'SSE';
    return 'Polling';
  });

  /** Emits raw incoming payloads from any transport (socket object or SSE string). */
  readonly message$ = new Subject<string | object>();

  /**
   * Emits whenever a transport successfully connects or reconnects.
   * Consumers should use this to trigger foreground sync (pull messages, recover logs, etc.).
   */
  readonly connected$ = new Subject<void>();

  /** Emits on each polling interval tick so the consumer can pull messages. */
  readonly pollTick$ = new Subject<void>();

  private socket: RealtimeSocket | null = null;
  private socketConnected = false;
  private socketConnecting = false;
  private stream: EventSource | null = null;
  private pollTimer: ReturnType<typeof setInterval> | null = null;
  private reconnectTimer: ReturnType<typeof setTimeout> | null = null;
  private socketReconnectTimer: ReturnType<typeof setTimeout> | null = null;
  private socketSseFallbackTimer: ReturnType<typeof setTimeout> | null = null;
  private shuttingDown = false;
  private socketConsecutiveFailures = 0;
  private socketDisabledUntil = 0;

  private activeUser: string | null = null;
  private networkReachableFn: () => boolean = () => true;

  constructor(private readonly api: ChatApiService) {}

  /**
   * Start the transport state machine for the given user.
   * Attempts Socket.io first, then falls back to SSE, with automatic reconnect.
   *
   * @param user          Normalized user identifier.
   * @param isNetworkReachable  Callback that returns whether the network is currently reachable.
   */
  connect(user: string, isNetworkReachable: () => boolean): void {
    this.disconnect();
    this.activeUser = user;
    this.networkReachableFn = isNetworkReachable;
    if (!this.networkReachableFn()) {
      return;
    }
    void this.connectSocketPreferred(user);
  }

  /**
   * Emit an event through the active socket and wait for an acknowledgement.
   * Returns `null` when the socket is not connected or the ack times out.
   */
  emitWithAck(
    eventName: string,
    payload: unknown,
    timeoutMs = SOCKET_ACK_TIMEOUT_MS
  ): Promise<Record<string, unknown> | null> {
    if (!this.socket || !this.socketConnected) {
      return Promise.resolve(null);
    }

    return new Promise<Record<string, unknown> | null>((resolve) => {
      let settled = false;
      const done = (value: Record<string, unknown> | null): void => {
        if (settled) return;
        settled = true;
        clearTimeout(timeoutId);
        resolve(value);
      };
      const timeoutId = setTimeout(() => done(null), timeoutMs);
      try {
        this.socket?.emit(eventName, payload, (ackPayload: unknown) => {
          if (ackPayload && typeof ackPayload === 'object') {
            done(ackPayload as Record<string, unknown>);
            return;
          }
          done(null);
        });
      } catch {
        done(null);
      }
    });
  }

  /** Tear down all transports and timers. */
  disconnect(): void {
    this.shuttingDown = true;
    this.stopSocketOnly();
    this.stopStreamOnly();

    if (this.pollTimer) {
      clearInterval(this.pollTimer);
      this.pollTimer = null;
    }
    if (this.reconnectTimer) {
      clearTimeout(this.reconnectTimer);
      this.reconnectTimer = null;
    }
    if (this.socketReconnectTimer) {
      clearTimeout(this.socketReconnectTimer);
      this.socketReconnectTimer = null;
    }
    if (this.socketSseFallbackTimer) {
      clearTimeout(this.socketSseFallbackTimer);
      this.socketSseFallbackTimer = null;
    }

    this.setTransportMode('polling');
    this.shuttingDown = false;
  }

  /** Start interval-based polling (dead-code path kept for completeness). */
  startPolling(user: string): void {
    if (this.pollTimer) {
      clearInterval(this.pollTimer);
    }

    this.pollTimer = setInterval(() => {
      this.pollTick$.next();
    }, POLL_INTERVAL_MS);

    if (!this.socketConnected && !this.stream) {
      this.setTransportMode('polling');
    }

    this.pollTick$.next();
    this.activeUser = user;
  }

  // ---------------------------------------------------------------------------
  //  Socket.io
  // ---------------------------------------------------------------------------

  private async connectSocketPreferred(user: string): Promise<void> {
    this.shuttingDown = false;
    if (this.socketDisabledUntil > Date.now()) {
      this.startSseFallback(user);
      this.scheduleSocketReconnect(user);
      return;
    }
    if (!this.networkReachableFn()) {
      this.startSseFallback(user);
      return;
    }
    if (this.socketConnecting) {
      return;
    }
    this.socketConnecting = true;

    if (this.socketSseFallbackTimer) {
      clearTimeout(this.socketSseFallbackTimer);
      this.socketSseFallbackTimer = null;
    }

    this.socketSseFallbackTimer = setTimeout(() => {
      this.socketSseFallbackTimer = null;
      if (!this.socketConnected && this.activeUser === user) {
        this.startSseFallback(user);
      }
    }, SOCKET_FALLBACK_TO_SSE_DELAY_MS);

    try {
      const socket = await this.api.createRealtimeSocket(user);
      if (this.activeUser !== user) {
        socket.disconnect();
        return;
      }
      this.shuttingDown = true;
      this.stopSocketOnly();
      this.shuttingDown = false;
      this.socket = socket;

      socket.on('connect', () => {
        if (this.activeUser !== user) return;
        this.resetSocketFailureState();
        this.socketConnected = true;
        this.socketConnecting = false;
        this.setTransportMode('socket');
        if (this.socketSseFallbackTimer) {
          clearTimeout(this.socketSseFallbackTimer);
          this.socketSseFallbackTimer = null;
        }
        this.stopStreamOnly();
        this.connected$.next();
      });

      socket.on('chat:message', (incoming: unknown) => {
        if (!incoming || typeof incoming !== 'object') return;
        this.message$.next(incoming as object);
      });

      socket.on('chat:connected', () => {
        if (this.activeUser !== user) return;
        this.resetSocketFailureState();
        this.socketConnected = true;
        this.setTransportMode('socket');
        this.stopStreamOnly();
        this.connected$.next();
      });

      socket.on('disconnect', () => {
        if (this.shuttingDown || this.activeUser !== user) return;
        this.shuttingDown = true;
        this.stopSocketOnly();
        this.shuttingDown = false;
        this.socketConnected = false;
        this.socketConnecting = false;
        this.setTransportMode('polling');
        this.handleSocketConnectFailure(user);
      });

      socket.on('connect_error', () => {
        if (this.shuttingDown || this.activeUser !== user) return;
        this.shuttingDown = true;
        this.stopSocketOnly();
        this.shuttingDown = false;
        this.socketConnected = false;
        this.socketConnecting = false;
        this.setTransportMode('polling');
        this.handleSocketConnectFailure(user);
      });

      socket.connect();
    } catch {
      this.shuttingDown = true;
      this.stopSocketOnly();
      this.shuttingDown = false;
      this.socketConnecting = false;
      this.socketConnected = false;
      this.setTransportMode('polling');
      this.handleSocketConnectFailure(user);
    }
  }

  // ---------------------------------------------------------------------------
  //  SSE
  // ---------------------------------------------------------------------------

  private startSseFallback(user: string): void {
    if (this.socketConnected || !this.networkReachableFn()) {
      return;
    }
    if (this.stream) {
      return;
    }
    try {
      this.stream = this.api.createMessageStream(user);
      this.setTransportMode('sse');
      this.stream.addEventListener('message', (event: MessageEvent<string>) => {
        this.message$.next(event.data);
      });
      this.stream.addEventListener('connected', () => {
        this.connected$.next();
      });
      this.stream.onerror = () => {
        this.stopStreamOnly();
        this.scheduleStreamReconnect(user);
      };
    } catch {
      this.scheduleStreamReconnect(user);
    }
  }

  // ---------------------------------------------------------------------------
  //  Failure handling & reconnect
  // ---------------------------------------------------------------------------

  private handleSocketConnectFailure(user: string): void {
    if (this.activeUser !== user) {
      return;
    }
    this.socketConsecutiveFailures += 1;
    this.startSseFallback(user);

    if (this.socketConsecutiveFailures >= SOCKET_MAX_FAILURES_BEFORE_COOLDOWN) {
      this.socketConsecutiveFailures = 0;
      this.socketDisabledUntil = Date.now() + SOCKET_FAILURE_COOLDOWN_MS;
    }

    this.scheduleSocketReconnect(user);
  }

  private resetSocketFailureState(): void {
    this.socketConsecutiveFailures = 0;
    this.socketDisabledUntil = 0;
  }

  private scheduleStreamReconnect(user: string): void {
    if (this.socketConnected || this.reconnectTimer) return;
    this.reconnectTimer = setTimeout(() => {
      this.reconnectTimer = null;
      if (this.activeUser !== user) return;
      this.startSseFallback(user);
    }, STREAM_RETRY_MS);
  }

  private scheduleSocketReconnect(user: string): void {
    if (this.socketReconnectTimer) return;
    const waitMs = this.socketDisabledUntil > Date.now()
      ? Math.max(SOCKET_RETRY_MS, this.socketDisabledUntil - Date.now())
      : SOCKET_RETRY_MS;
    this.socketReconnectTimer = setTimeout(() => {
      this.socketReconnectTimer = null;
      if (this.activeUser !== user || !this.networkReachableFn()) return;
      void this.connectSocketPreferred(user);
    }, waitMs);
  }

  // ---------------------------------------------------------------------------
  //  Internal teardown helpers
  // ---------------------------------------------------------------------------

  private stopSocketOnly(): void {
    this.socketConnected = false;
    this.socketConnecting = false;
    if (this.socket) {
      this.socket.disconnect();
      this.socket = null;
    }
    if (this.stream) {
      this.setTransportMode('sse');
    } else {
      this.setTransportMode('polling');
    }
  }

  private stopStreamOnly(): void {
    if (this.stream) {
      this.stream.close();
      this.stream = null;
    }
    if (!this.socketConnected) {
      this.setTransportMode('polling');
    }
  }

  private setTransportMode(mode: RealtimeTransportMode): void {
    if (this.transportMode() === mode) {
      return;
    }
    this.transportMode.set(mode);
  }
}

File: flutter_app/lib/core/realtime/realtime_transport_service.dart
/// Realtime transport service for WebSocket, SSE, and polling fallback.
///
/// This mirrors the Angular RealtimeTransportService behavior with the
/// same fallback strategy: Socket → SSE → Polling
library;

import 'dart:async';
import 'dart:convert';

import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:logger/logger.dart';
import 'package:socket_io_client/socket_io_client.dart' as socket_io;
import 'package:http/http.dart' as http;
import 'package:connectivity_plus/connectivity_plus.dart';

import '../config/app_config.dart';
import '../config/environment.dart';
import '../models/chat_models.dart';

final _logger = Logger(
  printer: PrettyPrinter(methodCount: 0, errorMethodCount: 5, lineLength: 80),
);

/// Transport mode enumeration
enum RealtimeTransportMode { socket, sse, polling }

/// Provider for the realtime transport service
final realtimeTransportServiceProvider = Provider<RealtimeTransportService>((ref) {
  return RealtimeTransportService();
});

/// Realtime transport service with fallback strategy
class RealtimeTransportService {
  /// Current transport mode
  RealtimeTransportMode _transportMode = RealtimeTransportMode.polling;
  RealtimeTransportMode get transportMode => _transportMode;

  /// Stream controller for incoming messages
  final _messageController = StreamController<dynamic>.broadcast();
  Stream<dynamic> get messageStream => _messageController.stream;

  /// Stream controller for connection events
  final _connectedController = StreamController<void>.broadcast();
  Stream<void> get connectedStream => _connectedController.stream;

  /// Stream controller for poll ticks
  final _pollTickController = StreamController<void>.broadcast();
  Stream<void> get pollTickStream => _pollTickController.stream;

  // Internal state
  socket_io.Socket? _socket;
  bool _socketConnected = false;
  bool _socketConnecting = false;
  http.Client? _sseClient;
  StreamSubscription? _sseSubscription;
  Timer? _pollTimer;
  Timer? _reconnectTimer;
  Timer? _socketReconnectTimer;
  Timer? _socketSseFallbackTimer;
  bool _shuttingDown = false;
  int _socketConsecutiveFailures = 0;
  int _socketDisabledUntil = 0;

  String? _activeUser;
  bool Function()? _isNetworkReachable;

  /// Human-readable label for current transport mode
  String get transportLabel {
    switch (_transportMode) {
      case RealtimeTransportMode.socket:
        return 'Socket';
      case RealtimeTransportMode.sse:
        return 'SSE';
      case RealtimeTransportMode.polling:
        return 'Polling';
    }
  }

  /// Start the transport state machine for the given user.
  /// Attempts Socket.io first, then falls back to SSE, with automatic reconnect.
  void connect(String user, {bool Function()? isNetworkReachable}) {
    disconnect();
    _activeUser = user;
    _isNetworkReachable = isNetworkReachable ?? () => true;

    if (!_isNetworkReachable!()) {
      return;
    }

    _connectSocketPreferred(user);
  }

  /// Emit an event through the active socket and wait for acknowledgement.
  /// Returns null when socket is not connected or ack times out.
  Future<Map<String, dynamic>?> emitWithAck(
    String eventName,
    dynamic payload, {
    Duration timeout = RealtimeConfig.socketAckTimeout,
  }) async {
    if (_socket == null || !_socketConnected) {
      return null;
    }

    final completer = Completer<Map<String, dynamic>?>();
    Timer? timeoutTimer;

    void done(Map<String, dynamic>? value) {
      if (!completer.isCompleted) {
        timeoutTimer?.cancel();
        completer.complete(value);
      }
    }

    timeoutTimer = Timer(timeout, () => done(null));

    try {
      _socket!.emitWithAck(eventName, payload, ack: (ackPayload) {
        if (ackPayload is Map<String, dynamic>) {
          done(ackPayload);
        } else {
          done(null);
        }
      });
    } catch (e) {
      done(null);
    }

    return completer.future;
  }

  /// Tear down all transports and timers
  void disconnect() {
    _shuttingDown = true;
    _stopSocketOnly();
    _stopSseOnly();

    _pollTimer?.cancel();
    _pollTimer = null;
    _reconnectTimer?.cancel();
    _reconnectTimer = null;
    _socketReconnectTimer?.cancel();
    _socketReconnectTimer = null;
    _socketSseFallbackTimer?.cancel();
    _socketSseFallbackTimer = null;

    _setTransportMode(RealtimeTransportMode.polling);
    _shuttingDown = false;
  }

  /// Start interval-based polling
  void startPolling(String user) {
    _pollTimer?.cancel();

    _pollTimer = Timer.periodic(RealtimeConfig.pollInterval, (_) {
      _pollTickController.add(null);
    });

    if (!_socketConnected && _sseSubscription == null) {
      _setTransportMode(RealtimeTransportMode.polling);
    }

    _pollTickController.add(null);
    _activeUser = user;
  }

  /// Dispose resources
  void dispose() {
    disconnect();
    _messageController.close();
    _connectedController.close();
    _pollTickController.close();
  }

  // ---------------------------------------------------------------------------
  // Socket.io Connection
  // ---------------------------------------------------------------------------

  Future<void> _connectSocketPreferred(String user) async {
    _shuttingDown = false;

    if (_socketDisabledUntil > DateTime.now().millisecondsSinceEpoch) {
      _startSseFallback(user);
      _scheduleSocketReconnect(user);
      return;
    }

    if (_isNetworkReachable != null && !_isNetworkReachable!()) {
      _startSseFallback(user);
      return;
    }

    if (_socketConnecting) {
      return;
    }
    _socketConnecting = true;

    _socketSseFallbackTimer?.cancel();
    _socketSseFallbackTimer = Timer(RealtimeConfig.socketFallbackToSseDelay, () {
      _socketSseFallbackTimer = null;
      if (!_socketConnected && _activeUser == user) {
        _startSseFallback(user);
      }
    });

    try {
      final socket = _createSocket(user);
      if (_activeUser != user) {
        socket.disconnect();
        return;
      }

      _shuttingDown = true;
      _stopSocketOnly();
      _shuttingDown = false;
      _socket = socket;

      socket.onConnect((_) {
        if (_activeUser != user) return;
        _resetSocketFailureState();
        _socketConnected = true;
        _socketConnecting = false;
        _setTransportMode(RealtimeTransportMode.socket);
        _socketSseFallbackTimer?.cancel();
        _socketSseFallbackTimer = null;
        _stopSseOnly();
        _connectedController.add(null);
        _logger.d('Socket connected');
      });

      socket.on('chat:message', (incoming) {
        if (incoming == null) return;
        _messageController.add(incoming);
      });

      socket.on('chat:connected', (_) {
        if (_activeUser != user) return;
        _resetSocketFailureState();
        _socketConnected = true;
        _setTransportMode(RealtimeTransportMode.socket);
        _stopSseOnly();
        _connectedController.add(null);
      });

      socket.onDisconnect((_) {
        if (_shuttingDown || _activeUser != user) return;
        _shuttingDown = true;
        _stopSocketOnly();
        _shuttingDown = false;
        _socketConnected = false;
        _socketConnecting = false;
        _setTransportMode(RealtimeTransportMode.polling);
        _handleSocketConnectFailure(user);
        _logger.d('Socket disconnected');
      });

      socket.onConnectError((error) {
        if (_shuttingDown || _activeUser != user) return;
        _shuttingDown = true;
        _stopSocketOnly();
        _shuttingDown = false;
        _socketConnected = false;
        _socketConnecting = false;
        _setTransportMode(RealtimeTransportMode.polling);
        _handleSocketConnectFailure(user);
        _logger.e('Socket connect error: $error');
      });

      socket.connect();
    } catch (e) {
      _shuttingDown = true;
      _stopSocketOnly();
      _shuttingDown = false;
      _socketConnecting = false;
      _socketConnected = false;
      _setTransportMode(RealtimeTransportMode.polling);
      _handleSocketConnectFailure(user);
      _logger.e('Socket creation error: $e');
    }
  }

  socket_io.Socket _createSocket(String user) {
    final serverUrl = Env.current.baseUrl.replaceAll('/notify', '');

    return socket_io.io(
      serverUrl,
      socket_io.OptionBuilder()
          .setPath(Env.current.socketPath)
          .setTransports(['polling', 'websocket'])
          .disableReconnection()
          .setAuth({'user': user})
          .setQuery({'user': user})
          .disableAutoConnect()
          .build(),
    );
  }

  // ---------------------------------------------------------------------------
  // SSE Fallback
  // ---------------------------------------------------------------------------

  void _startSseFallback(String user) {
    if (_socketConnected || (_isNetworkReachable != null && !_isNetworkReachable!())) {
      return;
    }
    if (_sseSubscription != null) {
      return;
    }

    try {
      final streamUrl = '${Env.current.baseUrl}${ApiEndpoints.stream}?user=${Uri.encodeComponent(user)}';
      _sseClient = http.Client();
      
      _setTransportMode(RealtimeTransportMode.sse);
      _logger.d('Starting SSE connection');

      // Note: For production, use a proper SSE client library
      // This is a simplified implementation
      _startSseStream(streamUrl, user);
    } catch (e) {
      _scheduleStreamReconnect(user);
      _logger.e('SSE start error: $e');
    }
  }

  void _startSseStream(String url, String user) async {
    try {
      final request = http.Request('GET', Uri.parse(url));
      request.headers['Accept'] = 'text/event-stream';
      request.headers['Cache-Control'] = 'no-cache';

      final response = await _sseClient!.send(request);
      
      if (response.statusCode != 200) {
        _stopSseOnly();
        _scheduleStreamReconnect(user);
        return;
      }

      _connectedController.add(null);

      _sseSubscription = response.stream
          .transform(utf8.decoder)
          .transform(const LineSplitter())
          .listen(
        (line) {
          if (line.startsWith('data: ')) {
            final data = line.substring(6);
            try {
              final parsed = jsonDecode(data);
              _messageController.add(parsed);
            } catch (e) {
              _messageController.add(data);
            }
          }
        },
        onError: (error) {
          _stopSseOnly();
          _scheduleStreamReconnect(user);
        },
        onDone: () {
          _stopSseOnly();
          _scheduleStreamReconnect(user);
        },
      );
    } catch (e) {
      _stopSseOnly();
      _scheduleStreamReconnect(user);
    }
  }

  // ---------------------------------------------------------------------------
  // Failure Handling & Reconnect
  // ---------------------------------------------------------------------------

  void _handleSocketConnectFailure(String user) {
    if (_activeUser != user) return;

    _socketConsecutiveFailures += 1;
    _startSseFallback(user);

    if (_socketConsecutiveFailures >= RealtimeConfig.maxSocketFailuresBeforeCooldown) {
      _socketConsecutiveFailures = 0;
      _socketDisabledUntil =
          DateTime.now().millisecondsSinceEpoch + RealtimeConfig.socketFailureCooldown.inMilliseconds;
    }

    _scheduleSocketReconnect(user);
  }

  void _resetSocketFailureState() {
    _socketConsecutiveFailures = 0;
    _socketDisabledUntil = 0;
  }

  void _scheduleStreamReconnect(String user) {
    if (_socketConnected || _reconnectTimer != null) return;

    _reconnectTimer = Timer(RealtimeConfig.streamRetryDelay, () {
      _reconnectTimer = null;
      if (_activeUser != user) return;
      _startSseFallback(user);
    });
  }

  void _scheduleSocketReconnect(String user) {
    if (_socketReconnectTimer != null) return;

    final now = DateTime.now().millisecondsSinceEpoch;
    final waitMs = _socketDisabledUntil > now
        ? (_socketDisabledUntil - now).clamp(RealtimeConfig.socketRetryDelay.inMilliseconds, double.infinity).toInt()
        : RealtimeConfig.socketRetryDelay.inMilliseconds;

    _socketReconnectTimer = Timer(Duration(milliseconds: waitMs), () {
      _socketReconnectTimer = null;
      if (_activeUser != user || (_isNetworkReachable != null && !_isNetworkReachable!())) return;
      _connectSocketPreferred(user);
    });
  }

  // ---------------------------------------------------------------------------
  // Internal Teardown Helpers
  // ---------------------------------------------------------------------------

  void _stopSocketOnly() {
    _socketConnected = false;
    _socketConnecting = false;
    if (_socket != null) {
      _socket!.disconnect();
      _socket!.dispose();
      _socket = null;
    }
    if (_sseSubscription != null) {
      _setTransportMode(RealtimeTransportMode.sse);
    } else {
      _setTransportMode(RealtimeTransportMode.polling);
    }
  }

  void _stopSseOnly() {
    _sseSubscription?.cancel();
    _sseSubscription = null;
    _sseClient?.close();
    _sseClient = null;
    if (!_socketConnected) {
      _setTransportMode(RealtimeTransportMode.polling);
    }
  }

  void _setTransportMode(RealtimeTransportMode mode) {
    if (_transportMode == mode) return;
    _transportMode = mode;
    _logger.d('Transport mode changed to: ${mode.name}');
  }
}

/// API endpoints for realtime connections
class ApiEndpoints {
  static const String stream = '/stream';
}

/// Network connectivity helper
class NetworkConnectivity {
  static Future<bool> isConnected() async {
    final connectivityResult = await Connectivity().checkConnectivity();
    return !connectivityResult.contains(ConnectivityResult.none);
  }

  static Stream<bool> get onConnectivityChanged {
    return Connectivity().onConnectivityChanged.map(
      (result) => !result.contains(ConnectivityResult.none),
    );
  }
}
